### PR TITLE
PB3

### PR DESCRIPTION
**Pull request by FedosovMax**
I've accomplished almost all tasks, but I haven't used UserRepository at all. I can't call it without @Autowire, so than I leave it and want to ask You about this meter. How can I call UserRepository methods in UserRepositoryRunner? Maybe You can tell me what I should learn to understand how it works. 
Is it all right that I made a PR without right solution?

### CODE REVIEW

File: src/main/java/com/mysite/entity/User.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
+package com.mysite.entity;
+
+import javax.persistence.*;
+
+@Entity
+@Table(name="user")
+public class User {
```
Comment (bohdanvan) [inline src/main/java/com/mysite/entity/User.java]: Add field `List<BookItem> bookItems` with appropriate annotations.

File: src/main/java/com/mysite/entity/com/mysite/repository/UserRepositoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+
+public class UserRepositoryImpl implements UserRepository {
+
+//    EntityManagerFactory entityManagerFactory = Persistence.createEntityManagerFactory(userPU);
+
+    private Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    @PersistenceContext
+    private EntityManager entityManager;
```
Comment (bohdanvan) [inline src/main/java/com/mysite/entity/com/mysite/repository/UserRepositoryImpl.java]: Add constructor for `EntityManager`.

File: src/main/java/com/mysite/entity/com/mysite/repository/UserRepositoryRunner.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    public static void main(String[] args) {
+
+
+        EntityManagerFactory emf = Persistence.createEntityManagerFactory("userPU");
+        EntityManager em = emf.createEntityManager();
+
+try{
+    User newUser = new User("Horn","password","Donald");
+
+    em.getTransaction().begin();
```
Comment (bohdanvan) [inline src/main/java/com/mysite/entity/com/mysite/repository/UserRepositoryRunner.java]: Fix indentations.

File: src/main/java/com/mysite/entity/com/mysite/repository/UserRepositoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import javax.persistence.Persistence;
+import javax.persistence.PersistenceContext;
+import javax.transaction.Transactional;
+
+
+public class UserRepositoryImpl implements UserRepository {
+
+//    EntityManagerFactory entityManagerFactory = Persistence.createEntityManagerFactory(userPU);
+
+    private Logger logger = LoggerFactory.getLogger(this.getClass());
```
Comment (bohdanvan) [inline src/main/java/com/mysite/entity/com/mysite/repository/UserRepositoryImpl.java]: The common practice is to use static field for logger.
`private static Logger logger = LoggerFactory.getLogger(UserRepositoryImpl.class);`

File: src/main/java/com/mysite/entity/com/mysite/repository/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package com.mysite.entity.com.mysite.repository;
```
Comment (bohdanvan) [inline src/main/java/com/mysite/entity/com/mysite/repository/UserRepository.java]: Strange package.

File: src/main/java/com/mysite/entity/com/mysite/repository/UserRepositoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
+package com.mysite.entity.com.mysite.repository;
```
Comment (bohdanvan) [inline src/main/java/com/mysite/entity/com/mysite/repository/UserRepositoryImpl.java]: Strange package.

**Comment by bohdanvan** [comment]
The PR is quite good, but with small issues.
Please, pass `EntityManager` through constructor of `UserRepositoryImpl` and write `Runner`, that uses this repository.

File: src/main/java/com/mysite/entity/User.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    private String login;
+
+    @Column(name="password")
+    private String password;
+
+    @Column(name="name")
+    private String name;
+
+    @OneToMany(mappedBy = "user", cascade = {CascadeType.DETACH, CascadeType.MERGE, CascadeType.PERSIST, CascadeType.REFRESH})
+    List<BookItem> bookItems;
```
Comment (bohdanvan) [inline src/main/java/com/mysite/entity/User.java]: Missed `private` access modifier.

Comment (bohdanvan) [inline src/main/java/com/mysite/entity/User.java]: Could we replace `cascade` with `CascadeType.ALL`?

File: src/main/java/com/mysite/repository/UserRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.Persistence;
+
+
+public class UserRepositoryImpl implements UserRepository {
+
+    EntityManagerFactory emf = Persistence.createEntityManagerFactory("userPU");
+    EntityManager em = emf.createEntityManager();
```
Comment (bohdanvan) [inline src/main/java/com/mysite/repository/UserRepositoryImpl.java]: Describe only field without initialization 
`private final EntityManager em`

File: src/main/java/com/mysite/repository/UserRepositoryRunner.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        em.getTransaction().begin();
+
+        User newUser = new User("Jonathan", "password", "Antuan");
+        newUser = theUserRepositoryImpl.save(newUser);  //Save
+        System.out.println(newUser + " was saved.");
+
+        User findUser = theUserRepositoryImpl.findById(1L); //FindById
+        System.out.println(findUser + " - it's your user.");
+
+        em.detach(findUser);
```
Comment (bohdanvan) [inline src/main/java/com/mysite/repository/UserRepositoryRunner.java]: Why do you use `detach`?

File: src/main/java/com/mysite/entity/User.java (hunk 7 above)
Comment (FedosovMax) [reply src/main/java/com/mysite/entity/User.java]: If I remember right, there are types that can cause something bad, but that variation is safe) In my opinion. I heard it in the course, but should google it.

File: src/main/java/com/mysite/repository/UserRepositoryImpl.java (hunk 8 above)
Comment (FedosovMax) [reply src/main/java/com/mysite/repository/UserRepositoryImpl.java]: Oh, understand, it means that I should call factory only from the Directory class, but I shouldn't have done it if runner.

File: src/main/java/com/mysite/repository/UserRepositoryRunner.java (hunk 9 above)
Comment (FedosovMax) [reply src/main/java/com/mysite/repository/UserRepositoryRunner.java]: I saw it in video example, it was in English, but author said that "merge" uses for merging detached objects. Is it wrong, I shouldn't have used it?

File: src/main/java/com/mysite/entity/User.java (hunk 7 above)
Comment (bohdanvan) [reply src/main/java/com/mysite/entity/User.java]: Ok, agree.

File: src/main/java/com/mysite/repository/UserRepositoryRunner.java (hunk 9 above)
Comment (bohdanvan) [reply src/main/java/com/mysite/repository/UserRepositoryRunner.java]: Try to merge without detaching and write, what will happen.

Comment (FedosovMax) [reply src/main/java/com/mysite/repository/UserRepositoryRunner.java]: It works too.

**Pull request merged**